cmd/promtool/rules.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/rules"
+	"github.com/prometheus/prometheus/tsdb/importer/blocks"
+)
+
+const blockSize = 2 // in hours
+
+// RuleImporter is the importer to backfill rules.
+type RuleImporter struct {
+	logger log.Logger
+	config RuleImporterConfig
+
+	groups      map[string]*rules.Group
+	groupLoader rules.GroupLoader
+
+	apiClient v1.API
+
+	writer *blocks.MultiWriter
+}
+
+// RuleImporterConfig is the config for the rule importer.
+type RuleImporterConfig struct {
+	Start        time.Time
+	End          time.Time
+	OutputDir    string
+	EvalInterval time.Duration
+	URL          string
+}
+
+// NewRuleImporter creates a new rule importer that can be used to backfill rules.
+func NewRuleImporter(logger log.Logger, config RuleImporterConfig) *RuleImporter {
+	return &RuleImporter{
+		config:      config,
+		groupLoader: rules.FileLoader{},
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client.
+func (importer *RuleImporter) Init() error {
+	importer.writer = blocks.NewMultiWriter(importer.logger,
+		importer.config.OutputDir,
+		importer.config.EvalInterval.Nanoseconds(),
+	)
+
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources.
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// LoadGroups reads groups from a list of rule files.
+func (importer *RuleImporter) LoadGroups(ctx context.Context, filenames []string) (errs []error) {
```

Comment:
This duplicate code is still here, let's not have logic like this in two places as it'll get out of sync.

Reply:
You left this same comment before and I replied to it here 
https://github.com/prometheus/prometheus/pull/7675#discussion_r487461258